### PR TITLE
Add ConsumerGroup scheduling

### PR DESCRIPTION
**Pull request by pierDipi**
Part of #1537 

## Proposed Changes

- Add ConsumerGroup scheduling

**Release Note**

```release-note
None
```

**Docs**

```
None
```

### CODE REVIEW

**Comment by pierDipi** [comment]
Flaky
/retest

**Comment by pierDipi** [comment]
Can I please get a review of this PR? @matzew @aliok @devguyio @aavarghese @lionelvillard

File: control-plane/pkg/apis/internals/kafka/eventing/v1alpha1/consumer_group_types.go
Comment (matzew) [inline control-plane/pkg/apis/internals/kafka/eventing/v1alpha1/consumer_group_types.go:56]: mind adding some comments on the new added APIs

Comment (pierDipi) [reply control-plane/pkg/apis/internals/kafka/eventing/v1alpha1/consumer_group_types.go:56]: Done, let me know if there are other places you meant.

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (matzew) [inline control-plane/pkg/reconciler/consumergroup/consumergroup.go:76]: `consumersByPlacement` -> `placementConsumers`?

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (matzew) [inline control-plane/pkg/reconciler/consumergroup/consumergroup.go:77]: when or why can this be `nil`?

Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go:77]: I'll add a comment, but it's when there is no placement for a consumer, so they need to be finalized.

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (matzew) [inline control-plane/pkg/reconciler/consumergroup/consumergroup.go:86]: can't we just pass in the `pc`, and stash them locally inside the `reconcileConsumersInPlacement` ?

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (matzew) [inline control-plane/pkg/reconciler/consumergroup/consumergroup.go:100]: if there is no consumer, we create/reconcile an internal consumer and return it  ?

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (matzew) [inline control-plane/pkg/reconciler/consumergroup/consumergroup.go:106]: I may lack context, but what state? the `IsLessThan` compares what?

**Comment by matzew** [comment]
I gave a first review. any more to read for better context ?

**Comment by pierDipi** [comment]
The scheduler doc and API may be helpful: https://github.com/knative/eventing/tree/main/pkg/scheduler

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go:86]: The idea was that in `reconcileConsumersInPlacement`, `pc.Placement` is guaranteed to be non-nil and that's clarified by passing a struct rather than a pointer to a struct.

Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go]: Done

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go:77]: Done

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go:109]: I'll make this part more explicit in a follow-up PR.

`IsLessThan` defines the order for deleting consumers in the same placement.

Imagine there are consumer `c1` and consumer `c2` in pod `p1`.
We want to have only one consumer for a placement, so we remove one of them based on their state.
In our case, we're considering readiness or the creation time as a state, meaning that we try to keep ready consumers over not ready ones or if they aren't ready we remove the most recently created ones.

File: control-plane/pkg/reconciler/consumergroup/consumergroup.go
Comment (pierDipi) [reply control-plane/pkg/reconciler/consumergroup/consumergroup.go:103]: Yes

**Comment by matzew** [comment]
/lgtm
thanks or addressing my comments. +1 on the follow up PR

/hold
waiting for final approval from @aavarghese and/or @lionelvillard

File: control-plane/pkg/reconciler/testing/objects_consumer.go
Comment (aavarghese) [inline control-plane/pkg/reconciler/testing/objects_consumer.go:90]: For consistency, can some of these funcs be renamed to `WithConsumerTopics`, ` WithConsumerPlacement`, `WithConsumerConfigs`, `WithConsumerBootstrapServersConfig`, etc. Wdyt?

**Review by aavarghese — APPROVED** [review]
All looks good! Func name change not critical for this iteration so..
/lgtm

**Comment by pierDipi** [comment]
Thanks!

/unhold

**Pull request merged**
